weather: Add new app with forecast

vkareh opened the pull request:
Added a new weather app that shows the current weather and forecast. If forecast is not available, it just shows the current weather.

![InfiniSim_2024-02-02_145908](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/d61651e4-06ca-48d5-9173-83586b620e7d)

Building the sim requires https://github.com/InfiniTimeOrg/InfiniSim/pull/138

FintasticMan commented:
Looks good! Is there enough space to add the ° symbol to the forecast? It looks a bit off without it IMO. It might also be nice to know the current min and max temp?

About the forecast not showing up with Breezy Weather + Gadgetbridge, could you check if it works with Tiny Weather Forecast Germany? The Gadgetbridge side should be working, so I suspect that Breezy Weather just doesn't provide that information.

FintasticMan commented on src/displayapp/screens/Weather.h:
```diff
@@ -0,0 +1,56 @@
+        void Refresh() override;
+
+      private:
+        const char* getCondition(const Pinetime::Controllers::SimpleWeatherService::Icons icon);
```
This function could be moved to displayapp/screens/WeatherSymbols.{h,cpp}, as I imagine it might be useful to other apps as well. Also, we use PascalCase for functions.

FintasticMan reviewed (changes requested):
One small thing, for the rest this looks good!

vkareh commented:
Ha! Found the issue. The wrong variable was being referenced in Gadgetbridge and so was not sending forecast data. I've submitted a patch to fix it: https://codeberg.org/Freeyourgadget/Gadgetbridge/pulls/3551

@FintasticMan Thanks for reviewing, I'll start addressing these issues sometime today.

FintasticMan commented:
Oh oopsie, that was my code in GB :sweat_smile:

Thanks for finding and submitting a fix.

FintasticMan commented:
I think that looks good, but I'm not exactly the best at UI design. If some more people chime in, I think that this is good! (You do still need to fix some formatting FYI. You can use the git pre-commit hook to do this automatically.)

vkareh commented:
@FintasticMan thanks - formatting is fixed now. The sim build breaks due to a new operator override in the Forecast struct. Should be fixed by https://github.com/InfiniTimeOrg/InfiniSim/pull/138

vkareh commented:
I removed the location display. It was causing my watch to crash and reboot. I might debug it later, but easier to remove it for now...

![InfiniSim_2024-02-01_112143](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/699fd0ff-744e-4f27-be5a-e34a187196d8)

FintasticMan commented:
Do positive numbers above 100 display fine? That is possible with Fahrenheit I think...

vkareh commented:
Yep, they look fine. The problem would be if it gets to 100 below zero (e.g. `-100` has 4 characters), but we have far bigger problems if that's the case :laughing: 

Acceptable display range would then be `-99` to `999`, from a purely technical point of view.

![InfiniSim_2024-02-01_113440](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/1af1afee-ea04-4bca-944f-142f1434d55a)

vkareh commented:
@minacode 
> One small thing: you could use InfiniTimes theme colors.

I'd be more than happy to take suggestions on how to improve the thing. Here are two options using the InfiniTimeTheme colors:

![InfiniSim_2024-02-01_130349](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/a925ef5a-9497-4ce7-8cef-42424d595ec8) ![InfiniSim_2024-02-01_130434](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/28e64a40-2cfb-43aa-8dfc-499452d1f652)

Any preference? Other color combinations? I haven't been able to figure out how to change the color of the text in the forecast table...

KaffeinatedKat commented:
that looks incredible, great work. Would it be possible to have the color change based on the temp? (eg. blue for cold and yellow/red for hot). That would really help bring it together imo

JustScott commented:
Wow this looks great! Do you think it would be possible to fit the date and/or week day above or below each forecast. Example: 'thu', 'fri', '2/1', '2/17'?

Oh and a quick suggestion for the colors: maybe have them dynamically change when they meet certain temp ranges, for example: blue when below freezing, orange when above freezing, and red when above some higher temp.

vkareh commented:
> that looks incredible, great work. Would it be possible to have the color change based on the temp? (eg. blue for cold and yellow/red for hot). That would really help bring it together imo

Thanks! I love this idea.

> Wow this looks great! Do you think it would be possible to fit the date and/or week day above or below each forecast. Example: 'thu', 'fri', '2/1', '2/17'?

I'll try, but I suspect there won't be much room for this. Also, this app is already annoyingly heavy (3388B as of last build), and calculating the days based on timestamps and then displaying that on a table will eat up a lot more, I think.

vkareh commented:
What are your thoughts on this? Anything 0°C or below is blue, 40°C or above is red, anything else is yellowish:
![InfiniSim_2024-02-01_132859](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/e54bb30a-5102-4f75-b8b1-2ff9c8d0e9f5) ![InfiniSim_2024-02-01_132904](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/ae9fc8fd-8ac0-4a85-9ccf-832a620826b6) ![InfiniSim_2024-02-01_133258](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/f4cf9de0-3a78-42ec-87da-4732c831625b)

JustScott commented:
> > Wow this looks great! Do you think it would be possible to fit the date and/or week day above or below each forecast. Example: 'thu', 'fri', '2/1', '2/17'?
> 
> I'll try, but I suspect there won't be much room for this. Also, this app is already annoyingly heavy (3388B as of last build), and calculating the days based on timestamps and then displaying that on a table will eat up a lot more, I think.

Perhaps just adding the weekday and date for today somewhere could work, like is done on the digital watch face where it's a less visible grey.

minacode commented:
Maybe increase the space between the big temperature and the description above?

vkareh commented:
How about now?
![InfiniSim_2024-02-01_154859](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/465e0a0a-e48a-440f-b2d6-fb7a60887aa0) ![InfiniSim_2024-02-01_155008](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/89374920-0116-48fb-b9dc-7f6c592ef28d)

I tried putting the current day/time (last updated time, really) but it was hard to align everything. It was just too cluttered...

KaffeinatedKat commented:
that looks really awesome, we don't really need to show the current day/time because that's easily accessible elsewhere. Having the foretasted days shown is more important. Great work

vkareh commented:
@MoonlightWave-12 
> The maximum should be at the top, because it is higher.

Good idea. Done:
![InfiniSim_2024-02-01_161858](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/4f2a2eb8-68e4-4b1b-9dfa-48e351defc29)

With regards to temperature colors, anything between 0°C and 40°C starts to become really subjective. Where I live currently 5°C is a very warm winter day, whereas 30°C is a hot summer day. However, I grew up in a place where 90°F is a mild winter day.

I'm okay with your ranges, but would like some consensus before comitting to those numbers. I also propose these:

* 0 or lower: blue (freezing)
* 0 - 4: cyan (ice warning)
* 4 - 32: yellow (nothing special, but color white doesn't look great here)
* 32 or higher: red (hot)

The 32°C number is from [wikipedia](https://en.wikipedia.org/wiki/Heat_index#Effects_of_the_heat_index_(shade_values)) but we can also use 27°C

vkareh commented:
@MoonlightWave-12 
> Maybe also colour the forecast-values, so that one can see the temperature-ranges easily at a glance?

That'd be nice, but I haven't figured out how to apply color to the table cells, only the table border (which I made black to blend in with the background)

FintasticMan commented:
Wow this is really coming together! I would suggest using 27 for hot rather than 32 personally, but I don't live anywhere very hot 😄. I'd say it'd be best to capitalise the first letter of the weekdays, as they are proper nouns, so it looks strange to me in lower-case.

vkareh commented:
@FintasticMan: done!

![InfiniSim_2024-02-01_172313](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/ef33018e-22da-4f86-b2e1-3d44e0a7875a) ![InfiniSim_2024-02-01_172333](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/52c9998e-d9dd-4906-a020-81083317f3d0)
![InfiniSim_2024-02-01_172319](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/6c4029ef-2eba-4fa6-8cd1-17a9f9378b95) ![InfiniSim_2024-02-01_173145](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/693fb3ef-fc33-4b32-b71c-6ec86be2557b)

vkareh commented:
I studied the lvgl table spec and was able to simplify the table code and re-color some table elements:
![InfiniSim_2024-02-02_140145](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/5d986071-e6b6-418b-bc3c-c7e7441c5f15)
I made them light gray for now but we can change them on an as-need basis.

KaffeinatedKat commented:
is it possible to color the table temps based on the same schema as the current? Or do they all have to be the same color. If you can change them all individually see what the color temps look like on the forecast. It'll either look great or not so great

vkareh commented:
@KaffeinatedKat - here's a selection with colorized forecast:
![InfiniSim_2024-02-02_145908](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/d61651e4-06ca-48d5-9173-83586b620e7d) ![InfiniSim_2024-02-02_145858](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/97e892b1-a4ab-4ecc-957c-0d3bfe02a7fb)
![InfiniSim_2024-02-02_145905](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/52cad742-0999-402e-8077-608a578b8552) ![InfiniSim_2024-02-02_145902](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/202c3e58-3651-4c65-9bb1-5e4379d81dc8)
![InfiniSim_2024-02-02_150153](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/00e24e88-4253-479c-83e8-6a607340a94d)

KaffeinatedKat commented:
I know that this feature is probably outside the scope for this PR, but when this is merged this would be the perfect app for the swipe left action on the clock

vkareh commented:
@KaffeinatedKat 
> I know that this feature is probably outside the scope for this PR, but when this is merged this would be the perfect app for the swipe left action on the clock

Last night I was thinking about a different feature. A setting where you can select a few apps (in a checkbox, for example) and those will be "always open" and you can cycle through them by swiping left. Also adding an auto-cycle option.

Definitely out of scope here. We can start a discussion thread about it outside this PR.

vkareh commented:
Bad news, the last commit (which colorizes the forecast temperatures) crashes my watch when loading the app... will debug and maybe revert depending on what I find.

vkareh commented:
Fixed the crash.
It was happening due to the task refresh callback being called every 20ms, which is probably not enough time for the fully-colorized table to render correctly before the next callback happened. I bumped that to 1000ms and it fixed the issue - there's no need to refresh the weather data that often anyway.

minacode commented:
What do you think about aligning the numbers per day?

Examples:
```
1
2

 1
12

 1
-3

  0
-11
```

vkareh commented:
@minacode

> What do you think about aligning the numbers per day?

not sure I understand this comment and the accompanying example

minacode commented:
I'm sorry, I thought about right-aligning the temperatures in the table.

vkareh commented:
@minacode 
> I'm sorry, I thought about right-aligning the temperatures in the table.

Good idea, I just pushed a new version with right-aligned temperatures. I think it looks good!

![InfiniSim_2024-02-05_092207](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/c0d553dd-09cd-47a1-9930-59244c418ee6)

minacode commented:
Yes! It's a little easier to the eyes to grasp the structure.

FintasticMan commented:
Is it always all the way right aligned? I.e. what will it do when you have a day with a 1-char value and one with a 2-char value?

vkareh commented:
> Is it always all the way right aligned? I.e. what will it do when you have a day with a 1-char value and one with a 2-char value?

Nothing different. The whole thing is an LVGL table with invisible borders, so it behaves like any numeric table (spreadsheet, etc.):
```
        +-----+-----+-----+-----+-----+-----+
right:  |    1|   -1|   11|  -11|  111| -111|
        +-----+-----+-----+-----+-----+-----+
center: |  1  |  -1 |  11 | -11 | 111 | -111|
        +-----+-----+-----+-----+-----+-----+
left:   |1    |-1   |11   |-11  |111  |-111 |
        +-----+-----+-----+-----+-----+-----+
```
The issue with anything other than right-aligned is that things look wrong, which is why spreadsheets typically default to right-aligning numbers.

This is probably a good visualization:
```
|right|centr|left |
+-----+-----+-----+
|    1|  1  |1    |
+-----+-----+-----+
|   -1|  -1 |-1   |
+-----+-----+-----+
|   11| 11  |11   |
+-----+-----+-----+
|  -11| -11 |-11  |
+-----+-----+-----+
|  111| 111 |111  |
+-----+-----+-----+
| -111| -111|-111 |
+-----+-----+-----+
```
Notice how the right-aligned column grows progressively from the right, and unit numbers are always aligned, as are the tens and hundreds, etc. Also the negative sign doesn't shift the number from alignment, as it's always away from the alignment.

FintasticMan commented:
It would've been nice to have it align right as far as its column is wide, if that makes any sense. But as I understand it that's not possible with LVGL tables.

I was thinking something like this:
```plaintext
+---+  +----+  +---+  +---+
| 5 |  | -1 |  |  5|  | -2|
| 1 |  | -5 |  |-23|  |-30|
+---+  +----+  +---+  +---+
```

FintasticMan commented:
I'm still not sure about the alignment. I think that right-aligning the numbers makes them easier to read, but I think it looks wrong in this situation:
![InfiniSim_2024-02-08_223348](https://github.com/InfiniTimeOrg/InfiniTime/assets/52415484/765ee2ee-175c-4042-8d59-f7e1772308b8)

Ideally, we'd have it look like this:
![InfiniSim_2024-02-08_223436](https://github.com/InfiniTimeOrg/InfiniTime/assets/52415484/15372d27-4118-44c2-a73c-d70eee7d451c) ![InfiniSim_2024-02-08_223504_edited](https://github.com/InfiniTimeOrg/InfiniTime/assets/52415484/7f094930-3277-4ce2-a1bc-fcdbfe3658cb) ![InfiniSim_2024-02-08_223608_edited](https://github.com/InfiniTimeOrg/InfiniTime/assets/52415484/43b57fb7-fb70-4616-852a-d2c4586958e7)

Is this possible? If not, we should decide whether right-aligned looks better in most cases or not.

minacode commented:
We could center the cells and pad them with spaces if necessary.

FintasticMan commented:
Good thinking!

vkareh commented:
Sounds like a good idea. I can try, but each cell will need to know how many digits are in the temperature above or below itself and then add padding based on that. Maybe just centering all of them is not that bad after all?

Anyway, I'll see what I can pull off next time I'm playing with this.

vkareh commented:
@FintasticMan @minacode - success! I managed to use character padding to align temperatures:
* With table borders (to visualize the alignment):
![InfiniSim_2024-02-10_131558](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/21b4d187-53ac-4cf8-bbac-bb1f0bd28293)
* Without table borders:
![InfiniSim_2024-02-10_132039](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/64f70b1f-c939-45a2-8296-3060a640a08e)
* With more realistic data:
![InfiniSim_2024-02-10_132314](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/b620e9f1-15ca-472f-8ad7-0d3d22698588)  ![InfiniSim_2024-02-10_132253](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/cdf316a2-a0e9-43dd-9db1-b692ca951060) ![InfiniSim_2024-02-10_132229](https://github.com/InfiniTimeOrg/InfiniTime/assets/259780/45c51764-c377-4f4a-b7ea-009b5842d30c)

FintasticMan commented on src/displayapp/screens/Weather.cpp:
```diff
@@ -0,0 +1,203 @@
+        minTemp = minTemp / 100 + (minTemp % 100 >= 50 ? 1 : 0);
+        lv_table_set_cell_value(forecast, 0, i, WeekDays[wday]);
+        lv_table_set_cell_value(forecast, 1, i, Symbols::GetSymbol(optCurrentForecast->days[i].iconId));
+        // Pad cells based on the largest number of digits on each column
+        std::string maxPadding = "";
+        std::string minPadding = "";
+        int maxSize = snprintf(nullptr, 0, "%d", maxTemp);
+        int minSize = snprintf(nullptr, 0, "%d", minTemp);
+        switch (maxSize - minSize) {
+          case -2:
+            maxPadding = "  ";
+            break;
+          case -1:
+            maxPadding = " ";
+            break;
+          case 1:
+            minPadding = " ";
+            break;
+          case 2:
+            minPadding = "  ";
+            break;
+        }
```
I don't really like using `std::string` here, as it can quite easily be done with a C array.

```suggestion
        // Pad cells based on the largest number of digits on each column
        char maxPadding[3] = "  ";
        char minPadding[3] = "  ";
        int maxSize = snprintf(nullptr, 0, "%d", maxTemp);
        int minSize = snprintf(nullptr, 0, "%d", minTemp);
        switch (maxSize - minSize) {
          case -1:
            maxPadding[1] = '\0';
            break;
          case 1:
            minPadding[1] = '\0';
            break;
        }
```

minacode replied:
This is missing
```cpp
  case 0:
    minPadding[0] = '\0';
    maxPadding[0] = '\0';
```

We also might want to handle longer numbers correctly, even though they are most likely wrong. This could be done by changing the `case 0` to `default`.

FintasticMan replied:
Oh yeah! Forgot about that

FintasticMan replied:
I've just updated it to handle larger numbers without having to add cases for them.

vkareh replied:
I love this, thanks!

minacode commented on src/displayapp/screens/Weather.h:
```diff
@@ -0,0 +1,58 @@
+
+      class Weather : public Screen {
+      public:
+        std::array<const char*, 7> WeekDays = {"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"};
```
1. There is something like this in the DatetimeController. Maybe you can use this instead so we don't have multiple definitions of the days names.
2. Why is this public?

vkareh replied:
Fixed.

JF002 commented on src/displayapp/screens/Weather.cpp:
```diff
@@ -0,0 +1,203 @@
+using namespace Pinetime::Applications::Screens;
+
+namespace {
+  lv_color_t temperatureColor(int16_t temperature) {
```
Please use PascalCase for function names

vkareh replied:
Fixed

JF002 commented on src/displayapp/screens/Weather.cpp:
```diff
@@ -0,0 +1,203 @@
+    return Colors::orange; // normal
+  }
+
+  uint8_t temperatureStyle(int16_t temperature) {
```
Please use PascalCase for function names

JF002 commented on src/displayapp/screens/Weather.cpp:
```diff
@@ -0,0 +1,203 @@
+        tempUnit = 'F';
+      }
+      temp = temp / 100 + (temp % 100 >= 50 ? 1 : 0);
+      maxTemp = maxTemp / 100 + (maxTemp % 100 >= 50 ? 1 : 0);
```
This calculation is done multiple times (and it's not easy to understand what it does - rounding?). Consider moving it to a dedicated function. The name of this function will also help understanding the meaning of it.

FintasticMan replied:
Yes, this is so that it gets rounded properly. I added it to the PTS watch face, but maybe we should add a function to SimpleWeatherService that does this calculation.

FintasticMan replied:
I've just created that function in #2015.

vkareh replied:
I added this function directly in the weather app to avoid having to push a new change to InfiniSim, but let me know if we should change this.

FintasticMan replied:
I think it would be better to have the function in the weather service tbh, but it doesn't matter so much.

JF002 reviewed (changes requested):
I'm very happy to see this new app since the protocol allows for forecast data, but we had no way of displaying those info so far!

This feature looks good overall and I guess we'll be able to merge it as soon as the remaining code change requests and review are addressed :)

FintasticMan commented on src/displayapp/screens/Weather.cpp:
```diff
@@ -0,0 +1,203 @@
+  if (currentForecast.IsUpdated()) {
+    auto optCurrentForecast = currentForecast.Get();
+    if (optCurrentForecast) {
+      std::tm localTime = *std::localtime((const time_t*) &optCurrentForecast->timestamp);
```
This cast should be done with `reinterpret_cast`.

```suggestion
      std::tm localTime = *std::localtime(reinterpret_cast<const time_t*>(&optCurrentForecast->timestamp));
```

vkareh replied:
Fixed

liamcharger commented:
Awesome work, this will make a great addition to the weather service when it is merged!

JF002 reviewed (approved):
LGTM! Thanks for your work @vkareh !

I've just tested it with Tiny Weather Forecast Germany and Gadgetbridge : 
![20240218_125006](https://github.com/InfiniTimeOrg/InfiniTime/assets/2261652/64784047-342f-43a8-a1c8-bbda7b618909)

Reminder : the current version of Gadgetbridge (0.78) has a bug that prevents forecast from being sent to the watch (https://codeberg.org/Freeyourgadget/Gadgetbridge/pulls/3551). The fix is merged, so I guess it'll be available in 0.79 (which is already tagged but not available in f-droid).
